File: Dynamic Programming/Coin_Change.java
public class Coin_Change {
    public int solveRecursive(int[] coins, int amount){
        if (amount == 0)
            return 0;
    
        if(amount < 0)
            return Integer.MAX_VALUE;

        int mini = Integer.MAX_VALUE;
        
        for (int j : coins) {
            int ans = solveRecursive(coins, amount - j);

            if(ans != Integer.MAX_VALUE){
                mini = Math.min(1+ans,mini);
            }
        }

        return mini;
    }
    public int solveMemorization(int[] coins, int amount,int[] dp){
        if (amount == 0)
            return 0;
    
        if(amount < 0)
            return Integer.MAX_VALUE;

        if(dp[amount] != -1)
            return dp[amount];

        int mini = Integer.MAX_VALUE;
        
        for (int j : coins) {
            int ans = solveRecursive(coins, amount - j);

            if(ans != Integer.MAX_VALUE){
                mini = Math.min(1+ans,mini);
            }
        }

        
        return dp[amount] = mini;
    }

    public int solveTabulation(int[] coins, int amount){

        int[] dp = new int[amount + 1];
        for (int i = 1 ; i < amount+1; i++) {dp[i] = Integer.MAX_VALUE;}
        
        dp[0] = 0;

        for(int i = 1 ; i <= amount ;i++){
            //i m trying to solve for evry amount 
            for(int j =0;j<coins.length;j++){
                if(i - coins[j] >= 0 && dp[i - coins[j]] != Integer.MAX_VALUE){
                    dp[i] = Math.min(dp[i], 1 + dp[i - coins[j]] );
                }
            }
        }

        if(dp[amount] == Integer.MAX_VALUE)
            return -1;

        return dp[amount];
    }

    public int coinChange(int[] coins, int amount) {

        // int ans = solveRecursive(coins,amount);

        // int[] dp = new int[amount + 1];
        // for (int i = 0 ; i < amount+1; i++) {dp[i] = -1;}
        // int ans = solveMemorization(coins,amount,dp);

        // if(ans == Integer.MAX_VALUE)
        //     return -1;

        // return ans;

        return solveTabulation(coins, amount);
    }
    public static void main(String[] args) {
        int[] coins = {1,2,5}; 
        int amount = 11;

        Coin_Change c = new Coin_Change();

        System.out.println(c.coinChange(coins, amount));
    }
}

File: Dynamic Programming/Coin_Change.cpp
// Minimum Elements
#include<iostream>
#include<vector>
#include<climits>
using namespace std;
int solveRecursive(vector<int> nums,int x){
    if(x == 0)
        return 0;
    if (x < 0)
        return INT_MAX;

    int mini = INT_MAX;
    for(auto i:nums){
        int ans = solveRecursive(nums,x - i);
        if (ans != INT_MAX)
            mini = min(mini,1 + ans);
    }
    return mini;
}
int solveMemorization(vector<int> nums,int x){

    vector<int> dp(x+1,INT_MAX);
    dp[0] = 0;

    for(int i = 1; i <= x ; i++ ){
        for (int j = 0; j < nums.size(); j++){
            if(i - nums[j] >= 0 && dp[i - nums[j]] != 0){
                dp[i] = min(dp[i],1 + dp[i - nums[j]]);
            }
        }
    }
    if(dp[x] == INT_MAX)
        return -1;
    
    return dp[x];

}



int minimumElements(vector<int> &num, int x){
    // Recursive call
    // int ans =  solveRecursive(num,x);
    // if(ans == INT_MAX)
    //     return -1;
    // return ans;

    return solveMemorization(num,x);
}
int main(){
    vector<int> nums = {1,2,5};
    int x = 7;
    cout<<solveRecursive(nums,x);
}